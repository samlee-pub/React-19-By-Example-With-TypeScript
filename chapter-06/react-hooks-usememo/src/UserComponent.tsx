/**
 * This component simulates a user component that receives props.
 * RegularUserComponent: Re-renders even when props don't change.
 * MemoizedUserComponent: Only re-renders when props change.
 */
import { memo } from "react";

// Regular component (re-renders even when props don't change)
export function RegularUserComponent({
  user,
  handleLogMessageEvent,
}: {
  user: { name: string; age: number };
  handleLogMessageEvent: (message: string) => void;
}) {
  console.log("RegularUser component rendered");
  handleLogMessageEvent("RegularUser component rendered");
  return (
    <div>{`Regular Component: {name: ${user.name}, age: ${user.age}}`}</div>
  );
}
// Memoized component (only re-renders when props change)
export const MemoizedUserComponent = memo(function MemoizedUser({
  user,
  handleLogMessageEvent,
}: {
  user: { name: string; age: number };
  handleLogMessageEvent: (message: string) => void;
}) {
  console.log("MemoizedUser component rendered");
  handleLogMessageEvent("MemoizedUser component rendered");
  return (
    <div>{`Memoized Component: {name: ${user.name}, age: ${user.age}}`}</div>
  );
});
