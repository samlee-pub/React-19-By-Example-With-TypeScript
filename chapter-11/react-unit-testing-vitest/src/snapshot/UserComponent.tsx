/**
 * This component demonstrates a simple user card that can be used for
 * Vitest snapshot testing. Snapshot testing ensures that the component's
 * output does not change unexpectedly over time.
 */
export function UserComponent(user: { name: string; email: string }) {
  return (
    <div className="user-card">
      <h2>${user.name}</h2>
      <p>${user.email}</p>
    </div>
  );
}
