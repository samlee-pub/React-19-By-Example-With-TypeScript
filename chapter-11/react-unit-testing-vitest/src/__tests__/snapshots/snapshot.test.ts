/**
 * Vitest Snapshot Testing Suite
 *
 * This test file demonstrates various snapshot testing techniques using Vitest.
 * It covers basic snapshots, inline snapshots, and partial snapshot matching
 * for different types of data and React components.
 *
 * Snapshot testing is ideal for:
 * - Preventing regression in UI components
 * - Testing complex data structures
 * - Ensuring serialized outputs remain consistent
 * - Catching unexpected changes in rendered output
 */
import { describe, test, expect } from "vitest";
import { UserComponent } from "../../snapshot/UserComponent";

/**
 * This suite explores different approaches to snapshot testing:
 * - Basic snapshots: For React components and complex objects
 * - Inline snapshots: With snapshots embedded in test code
 * - Partial matching: Flexible snapshots with dynamic values
 */
describe("Snapshots", () => {
  describe("basic snapshot test", () => {
    test("test on setup", () => {
      const user = { name: "John Doe", email: "john@example.com" };
      const result = UserComponent(user);
      // Creates/compares snapshot
      expect(result).toMatchSnapshot();
    });

    test("inline snapshot", () => {
      const data = { items: ["apple", "banana"], count: 2 };
      expect(data).toMatchSnapshot();
      expect(data).toMatchInlineSnapshot(`
        {
          "count": 2,
          "items": [
            "apple",
            "banana",
          ],
        }
      `);
    });
    test("test snapshot match", () => {
      const data = { items: ["apple", "banana"], count: Math.random() };
      expect(data).toMatchSnapshot({
        count: expect.any(Number) as number,
      });
    });
  });
});
