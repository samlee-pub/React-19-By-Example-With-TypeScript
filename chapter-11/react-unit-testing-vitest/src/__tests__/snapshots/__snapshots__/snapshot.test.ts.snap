// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshots > basic snapshot test > inline snapshot 1`] = `
{
  "count": 2,
  "items": [
    "apple",
    "banana",
  ],
}
`;

exports[`Snapshots > basic snapshot test > test on setup 1`] = `
<div
  className="user-card"
>
  <h2>
    $
    John Doe
  </h2>
  <p>
    $
    john@example.com
  </p>
</div>
`;

exports[`Snapshots > basic snapshot test > test snapshot match 1`] = `
{
  "count": Any<Number>,
  "items": [
    "apple",
    "banana",
  ],
}
`;
